feat: add performance tables

MrtinoRG opened the pull request:
## Summary by Sourcery

Add scripts to generate and collect performance tables for various models, including human-scored questions, and produce LaTeX tables for performance metrics.

New Features:
- Introduce scripts to generate performance tables from model scores, including human-scored questions.

Enhancements:
- Add functionality to collect and process performance scores for various models, including human baseline comparisons.

Documentation:
- Generate LaTeX tables for performance metrics, including calculation, knowledge, reasoning, intuition, and difficulty levels.

MrtinoRG commented:
![{735303D4-4E9D-4DDF-AC18-2019D7435950}](https://github.com/user-attachments/assets/24f61677-d3d8-464c-b313-79fdf7fa52c4)

This is the performance table for the entire dataset

![{BC13D05E-0427-49A0-AC4D-55F3327D6AC0}](https://github.com/user-attachments/assets/c146cdec-5a7c-4405-a4c6-bf1e8b164979)

And this is the one for the human subset.

Tell me if you think something should be changed

AdrianM0 commented:
I like it!

kjappelbaum commented on src/scripts/collect_performance_table_scores.py:
```diff
@@ -0,0 +1,258 @@
+
+if __name__ == "__main__":
+    chembench_repo = obtain_chembench_repo()
+    reports = glob(
+        os.path.join(chembench_repo, "reports",
+                     "**", "reports", "**", "*.json")
+    )
+    overall_files = glob(
+        os.path.join(chembench_repo, "reports", "**", "*.json")
+    )
+
+    models = list(set([Path(p).parent for p in reports]))
```
cool! in principle you can also always rely on `data/ model_score_dicts.pkl` 

here is an example https://www.dropbox.com/scl/fi/aszdlklymc6q0ec3z7ftz/model_score_dicts.pkl?rlkey=p4lv9gccqx8bhejgp3jaaspgp&dl=0

kjappelbaum replied:
but I can also adjust this later

kjappelbaum commented on src/scripts/collect_performance_table_scores.py:
```diff
@@ -0,0 +1,258 @@
+        os.mkdir(outpath)
+
+    datafolder = os.path.join(chembench_repo, "data")
+    human_baseline_folder = os.path.join(chembench_repo, "reports", "humans")
+
+    model_overall_scores = []
+    overall_scores = []
+    overall_human_scores = []
+    model_scores_human_subset = []
+    for i, file in enumerate(models):
+        try:
+            p = Path(file).parts[-3]
+            logger.info(f"Running for model {p}")
+            human_results = combine_scores_for_model(file.parent, datafolder, human_baseline_folder)
+            model_scores_human_subset.append({
+                "model": p,
+                "fraction_correct": human_results["fraction_correct"]
+            })
+            model_results = combine_scores_for_model(file.parent, datafolder)
+            model_overall_scores.append({
+                "model": p,
+                "fraction_correct": model_results["fraction_correct"]
+            })
+            overall_scores.append({
+                "model": p,
+                "scores": model_results,
+            })
+            overall_human_scores.append({
+                "model": p,
+                "scores": human_results,
+            })
+        except Exception as e:
+            logger.error(f"Error processing {file}: {e}")
```
here you'd have 

```
src/data/humans_as_models_scores_combined.pkl
src/data/humans_as_models_scores_no_tools.pkl
src/data/humans_as_models_scores_tools.pkl
```

kjappelbaum commented on src/scripts/make_performance_tables.py:
```diff
@@ -0,0 +1,299 @@
+
+    # Wrap the table with \resizebox
+    latex_table = (
+        r"\begin{table}[htbp]\centering" + "\n" +
```
I think I'd only put the `latex_table` in the file and then handle caption & co in the latex file

kjappelbaum commented on src/scripts/make_performance_tables.py:
```diff
@@ -0,0 +1,299 @@
+
+
+   # Function to bold the best value (assuming higher is better)
+    def bold_best(series):
+        best = series.astype(float).max()  # Modify as per your criteria (e.g., use min() for smaller is better)
+        return series.apply(lambda x: f"\\textbf{{{x}}}" if float(x) == best else x)
```
this is very cool! super nice that this bolding happens automatically!

MrtinoRG commented:
I have now removed all the additional processing of the `latex_table` remaining only the table itself (no caption or re-sizing).

Also, I added sorting so the table will be sorted alphabetically. I did not change how to collect the results because that little adjustments can be done afterwards

kjappelbaum commented:
thanks! I'll merge now and try to massage it into a good place

kjappelbaum commented:
I'm adding them into `Snakemake` at the moment. Then they will be always rerun if there is a change
